tsdb/db.go

```diff
@@ -1510,35 +1510,45 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
-				dir := filepath.Join(db.Dir(), uid.String())
-				if err := os.RemoveAll(dir); err != nil {
-					level.Error(db.logger).Log("msg", "failed to delete block after failed `CleanTombstones`", "dir", dir, "err", err)
-				}
-			}
-		}
-	}()
+	var (
+		cleanUpCompleted = false
+		blocks           []*Block
+	)
 
-	db.mtx.RLock()
-	blocks := db.blocks[:]
-	db.mtx.RUnlock()
+	// Repeat cleanup until there is no tombstones left
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
 
-	for _, b := range blocks {
-		if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
-			err = errors.Wrapf(er, "clean tombstones: %s", b.Dir())
-			return err
-		} else if uid != nil { // New block was created.
-			newUIDs = append(newUIDs, *uid)
+		blocks = db.Blocks()
+		for _, b := range blocks {
+			if b.tombstones.Total() > 0 {
+				cleanUpCompleted = false
+
+				fmt.Println("Trying to clean block: ", b.meta.ULID)
+
+				if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
+					return errors.Wrapf(er, "clean tombstones: %s", b.Dir())
+				} else if uid != nil { // New block was created.
+					fmt.Println("Tombstones cleaned for block: ", b.meta.ULID)
+					fmt.Println("New compacted block: ", uid)
+
+					if err = db.reload(); err != nil { // Will try to delete old block
+						// Delete directory of new block
+						dir := filepath.Join(db.Dir(), uid.String())
+						if err := os.RemoveAll(dir); err != nil {
+							fmt.Println("Failed to remove dir: ", dir, " err: ", err)
```

Comment:
Windows tests tell that the new block couldn't be deleted because it is opened somewhere. Couldn't figure out how to close it though